bundles/storage/org.eclipse.smarthome.storage.json/src/main/java/org/eclipse/smarthome/storage/json/internal/JsonStorage.java

```diff
@@ -295,49 +295,64 @@ public void commitDatabase() {
             writeDatabaseFile(new File(file.getParent() + File.separator + BACKUP_EXTENSION,
                     System.currentTimeMillis() + SEPARATOR + file.getName()), json);
 
+            cleanupBackups();
             deferredSince = 0;
         }
     }
 
+    private void cleanupBackups() {
+        List<Long> fileTimes = calculateFileTimes();
+
+        // delete the oldest
+        if (fileTimes.size() > maxBackupFiles) {
+            for (int counter = 0; counter < fileTimes.size() - maxBackupFiles; counter++) {
+                File deleter = new File(file.getParent() + File.separator + BACKUP_EXTENSION,
+                        fileTimes.get(counter) + SEPARATOR + file.getName());
+                deleter.delete();
+            }
+        }
+    }
+    
     private class CommitTimerTask extends TimerTask {
         @Override
         public void run() {
             // Save the database
             commitDatabase();
-            List<Long> fileTimes = calculateFileTimes();
-
-            // delete the oldest
-            if (fileTimes.size() > maxBackupFiles) {
-                for (int counter = 0; counter < fileTimes.size() - maxBackupFiles; counter++) {
-                    File deleter = new File(file.getParent() + File.separator + BACKUP_EXTENSION,
-                            fileTimes.get(counter) + SEPARATOR + file.getName());
-                    deleter.delete();
-                }
-            }
         }
     }
 
     public synchronized void deferredCommit() {
+        // Stop any existing timer
+        if (commitTimerTask != null) {
+            commitTimerTask.cancel();
+            commitTimerTask = null;
+        }
+        
         // Handle a maximum time for deferring the commit.
         // This stops a pathological loop preventing saving
-        if (deferredSince != 0 && deferredSince < System.nanoTime() - maxDeferredPeriod) {
+        if (deferredSince != 0 && deferredSince < System.nanoTime() - (maxDeferredPeriod * 1000L)) {
             commitDatabase();
+            // as we committed the database now, there is no need to start a new commit timer
+            return;
         }
+
         if (deferredSince == 0) {
             deferredSince = System.nanoTime();
         }
 
-        // Stop any existing timer
-        if (commitTimerTask != null) {
-            commitTimerTask.cancel();
-            commitTimerTask = null;
-        }
-
         // Create the timer task
         commitTimerTask = new CommitTimerTask();
 
         // Start the timer
         commitTimer.schedule(commitTimerTask, writeDelay);
     }
-
+    
+    public synchronized void deactivate() {
+        commitDatabase();
+        // Stop any existing timer
```

Comment:
Shouldn't an existing timer be always cleared within the the `commitDatabase()` method?  If this were the case, we would not need this additional method at all.